compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/EncodeV1.hs

```diff
@@ -84,28 +91,44 @@ encodeSourceLoc interned SourceLoc{..} =
         (fromIntegral slocEndCol)))
 
 encodePackageRef :: PackageRefCtx -> PackageRef -> Just P.PackageRef
-encodePackageRef interned = Just . \case
+encodePackageRef ctx = Just . \case
     PRSelf -> P.PackageRef $ Just $ P.PackageRefSumSelf P.Unit
     PRImport pkgid -> P.PackageRef $ Just $
       maybe (P.PackageRefSumPackageId $ encodePackageId pkgid)
             (P.PackageRefSumInternedId . fromIntegral)
-            $ interned pkgid
-
-internPackageRefIds :: Package -> (PackageRefCtx, [PackageId])
-internPackageRefIds pkg
-  | packageLfVersion pkg `supports` featureInternedPackageIds =
-      let set = S.fromList $ pkg ^.. packageRefs._PRImport
-          lookup pkgid = fromIntegral <$> pkgid `S.lookupIndex` set
-      in (lookup, S.toAscList set)
-  | otherwise = (const Nothing, [])
+            $ internPackageId ctx pkgid
+
+internPackageRefIds :: Package -> (PackageRefCtx, [PackageId], [ModuleName])
+internPackageRefIds pkg =
+  let (internPackageRef, packageIds) =
+        if packageLfVersion pkg `supports` featureInternedPackageIds
+        then let set = S.fromList $ pkg ^.. packageModuleRef._1._PRImport
+                 lookup pkgid = fromIntegral <$> pkgid `S.lookupIndex` set
+             in (lookup, S.toAscList set)
+        else (const Nothing, [])
+      (internModuleRef, moduleNames) =
+        if packageLfVersion pkg `supports` featureInternedModuleNames
+        then let freq = M.unionsWith (\_ _ -> True) $ pkg ^.. packageModuleRef._2.to (`M.singleton` False)
```

Comment:
I think this might be clearer if you use a `Map ModuleName Int` to track frequencies and filter on `> 0` or at least add a comment that points out what the `Bool` represents.